### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDR-Tool** [fcdl94/MiB#72]
Hi, In **MiB**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using

```
absl-py==0.8.0
apex==0.1
apturl==0.5.2
asn1crypto==0.24.0
astor==0.8.0
attrs==19.1.0
Automat==0.6.0
backcall==0.1.0
bleach==3.1.0
Brlapi==0.6.6
certifi==2018.1.18
chardet==3.0.4
click==6.7
colorama==0.3.7
command-not-found==0.3
configobj==5.0.6
constantly==15.1.0
cryptography==2.1.4
cupshelpers==1.0
cvxpy==1.0.25
cycler==0.10.0
decorator==4.4.0
defer==1.0.6
defusedxml==0.6.0
dill==0.3.1.1
distro-info===0.18ubuntu0.18.04.1
ecos==2.0.7.post1
entrypoints==0.3
future==0.17.1
gast==0.3.1
google-pasta==0.1.7
grpcio==1.23.0
h5py==2.10.0
httplib2==0.9.2
hyperlink==17.3.1
idna==2.6
incremental==16.10.1
inplace-abn==1.0.7
ipykernel==5.1.2
ipython==7.8.0
ipython-genutils==0.2.0
ipywidgets==7.5.1
jedi==0.15.1
Jinja2==2.10.1
joblib==0.11
jsonschema==3.0.2
jupyter==1.0.0
jupyter-client==5.3.3
jupyter-console==6.0.0
jupyter-core==4.5.0
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.0
keyring==10.6.0
keyrings.alt==3.0
kiwisolver==1.1.0
language-selector==0.1
launchpadlib==1.10.6
lazr.restfulclient==0.13.5
lazr.uri==1.0.3
louis==3.5.0
macaroonbakery==1.1.3
Mako==1.0.7
Markdown==3.1.1
MarkupSafe==1.1.1
matplotlib==3.1.1
mistune==0.8.4
multiprocess==0.70.9
nbconvert==5.6.0
nbformat==4.4.0
netifaces==0.10.4
nose==1.3.7
notebook==6.0.1
numpy==1.17.2
oauth==1.0.1
olefile==0.45.1
osqp==0.6.1
PAM==0.4.2
pandocfilters==1.4.2
parso==0.5.1
pexpect==4.7.0
pickleshare==0.7.5
Pillow==6.1.0
pluggy==0.6.0
prometheus-client==0.7.1
prompt-toolkit==2.0.9
protobuf==3.9.1
ptyprocess==0.6.0
py==1.5.2
pyasn1==0.4.2
pyasn1-modules==0.2.1
pycairo==1.16.2
pycrypto==2.6.1
pycups==1.9.73
Pygments==2.4.2
pygobject==3.26.1
pymacaroons==0.13.0
PyNaCl==1.1.2
pyOpenSSL==17.5.0
pyparsing==2.4.2
pyRFC3339==1.0
pyrsistent==0.15.4
pyserial==3.4
pytest==3.3.2
python-apt==1.6.4
python-dateutil==2.8.0
python-debian==0.1.32
pytz==2018.3
pyxdg==0.25
PyYAML==3.12
pyzmq==18.1.0
qtconsole==4.5.5
reportlab==3.4.0
requests==2.18.4
requests-unixsocket==0.1.5
scikit-learn==0.19.1
scipy==1.3.1
screen-resolution-extra==0.0.0
scs==2.1.1.post2
SecretStorage==2.3.1
Send2Trash==1.5.0
service-identity==16.0.0
simplegeneric==0.8.1
simplejson==3.13.2
six==1.12.0
ssh-import-id==5.7
system-service==0.3
systemd-python==234
tensorboard==1.14.0
tensorboardX==1.8
tensorflow-estimator==1.14.0
tensorflow-gpu==1.14.0
termcolor==1.1.0
terminado==0.8.2
testpath==0.4.2
torch==1.2.0
torchvision==0.4.0
tornado==6.0.3
traitlets==4.3.2
Twisted==17.9.0
ubuntu-drivers-common==0.0.0
ufw==0.36
urllib3==1.22
usb-creator==0.3.3
wadllib==1.3.2
wcwidth==0.1.7
webencodings==0.5.1
Werkzeug==0.15.6
widgetsnbextension==3.5.1
wrapt==1.11.2
xkit==0.0.0
zope.interface==4.3.2
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **multiprocess** can be changed to *==0.70.4*.
The version constraint of dependency **multiprocess** can be changed to *>=0.70.4,<=0.70.4*.
The version constraint of dependency **Pillow** can be changed to *==9.2.0*.
The version constraint of dependency **Pillow** can be changed to *>=2.0.0,<=9.1.1*.
The version constraint of dependency **pyasn1** can be changed to *>=0.4.1,<=0.4.8*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

<details><summary>The calling methods from the multiprocess</summary>
 <pre>logger.debug
logger.info
</pre>
</details>
<details><summary>The calling methods from the Pillow</summary>
 <pre>PIL.Image.open
</pre>
</details>
<details><summary>The calling methods from the pyasn1</summary>
 <pre>open
</pre>
</details>
<details><summary>The calling methods from the all methods</summary>
 <pre>torch.log_softmax
utils.logger.Logger.info
self._mean.reshape
self.n_classes.mask.label_pred.int.mask.label_true.astype.self.n_classes.np.bincount.reshape.sum
dict
dataset.transform.Resize
self.mod2
any
metrics.StreamSegMetrics
slice
range.keys
self.licarl.item
x.split
join
argparser.modify_command_options
model.named_parameters
torchvision.transforms.functional.adjust_contrast
self.book.clear
utils.loss.KnowledgeDistillationLoss
os.makedirs
self.lambd
torch.nn.Linear
p.grad.detach
torch.cat
metrics.update
p.grad.detach.pow
torch.nn.functional.binary_cross_entropy_with_logits
mask.float
torch.no_grad
utils.logger.Logger.add_image
self.mod5
torch.exp
NotImplementedError
VOCSegmentation
torch.randint
os.path.join
os.path.isdir
numpy.load
n.self.model_old_dict.p.detach.pow
model.eval
tbl.items
matplotlib.pyplot.subplots
type
torch.utils.data.DataLoader
targets.sum.loss.torch.masked_select.sum
RW
mask.label_true.astype
self.regularizer.load_state_dict
inputs.shape.labels_new.F.one_hot.float
train.Trainer
torch.nn.functional.pad
tasks.get_task_labels
self.fisher_old.items
n.self.score.to
random.random
numpy.random.choice
list.append
warnings.warn
random.uniform
Compose
images.detach.cpu.numpy
self.IncrementalSegmentationModule.super.__init__
key.self.fisher_old.to
denorm
self.convs
targets.sum.loss.torch.masked_select.mean
os.path.isfile
class_loss.torch.tensor.to
torch.nn.CrossEntropyLoss
torch.log
module.named_children
matplotlib.use
self.PolyLR.super.__init__
torch.optim.lr_scheduler.StepLR.state_dict
self.GlobalAvgPool2d.super.__init__
torch.distributed.get_rank
utils.Denormalize
self._fast_hist
p.clone
numpy.concatenate
utils.PolyLR
model.module.init_new_classifier
torch.nn.functional.pad.view
self.classes.torch.FloatTensor.torch.log.to
target.label2color.transpose.astype
loss.mean.mean
torch.nn.functional.one_hot
numpy.diag.sum
torchvision.transforms.functional.resized_crop
torch.nn.Conv2d.append
labels.cpu.numpy.cpu
self._global_pooling
self.lde_loss
utils.filter_images
torch.logsumexp.unsqueeze
apex.amp.scale_loss
modules.GlobalAvgPool2d
n.self.model_temp.to
math.sqrt
optim.zero_grad
torchvision.transforms.functional.adjust_saturation
idxs_path.np.load.tolist
logging.basicConfig
self.get_score
utils.loss.UnbiasedKnowledgeDistillationLoss
self.head
fig.tight_layout
torch.nn.functional.pad.repeat
p.clone.detach.cpu
self.regularizer.update
numpy.bincount
scaled_loss.backward
p.torch.clone.detach
argparse.ArgumentParser
t
torch.nn.functional.avg_pool2d
self.red_bn
torch.optim.lr_scheduler.StepLR.load_state_dict
torch.distributed.reduce
label2color
self.score.items
convert_bn2gn
train.Trainer.load_state_dict
transform
mod
epoch_loss.torch.tensor.to
cls.bias.data.copy_
images.to.detach
torch.nn.GroupNorm.add_module
self.DeeplabV3.super.__init__
self.model.named_parameters
m
samples.cpu.numpy
torch.nn.functional.nll_loss
self._transform_tag
torch.optim.SGD.load_state_dict
score.items
f.readlines
modules.DeeplabV3
logger.debug
torch.nn.functional.leaky_relu
utils.logger.Logger
idxs.append
opts.backbone.models.__dict__
self.convs.add_
apex.parallel.DistributedDataParallel.state_dict
inputs.size
norm_act
m.eval
torch.tensor
lt.flatten
numpy.random.seed
torch.softmax
x.dim
vars
int
AdeSegmentation
metrics.synch
apex.parallel.DistributedDataParallel.cuda
metrics.reset
apex.parallel.DistributedDataParallel.parameters
torchvision.transforms.functional.crop
inputs.shape.targets.shape.range.x.x.torch.tensor.to
__all__.append
apex.amp.initialize
n.self.model_old_dict.p.pow.n.self.score_actual.sum
sorted
utils.logger.Logger.add_results
torch.zeros_like
mat.max
n.self.model_old_dict.p.n.self.fisher_old.sum
tasks.get_task_list
logger.info
PIL.Image.open
p.clone.detach
self.ResNet.super.__init__
torch.from_numpy
torch.nn.functional.interpolate
torchvision.transforms.functional.rotate
self.body
math.log
dataset.transform.RandomResizedCrop
numpy.mean
lbl.label2color.transpose
self.logger.add_image
sample.apply_
TypeError
random.seed
round
fil
inputs.shape.labels_new.F.one_hot.float.permute
float
logging.info
modules.ResidualBlock
model_old.state_dict
tensorboardX.SummaryWriter
utils.logger.Logger.add_table
torch.nn.MSELoss
train_loader.sampler.set_epoch
prediction.cpu.numpy
math.exp
labels.cpu.numpy.to
lp.flatten
functools.partial
torch.nn.init.calculate_gain
voc_cmap
self.get_score.items
self.ResidualBlock.super.__init__
segmentation_module.make_model
inputs.narrow.narrow
x.idxs.append
self.order.index
self.lkd_loss
torch.nn.BCEWithLogitsLoss
labels.outputs.mean
train.Trainer.validate
torchvision.transforms.functional.pad
os.path.expanduser
labels.cpu.numpy
os.path.exists
self.get_params
setattr
torch.FloatTensor
get_dataset
IncrementalSegmentationModule
torch.masked_select
optim.step
repr
freq.iu.freq.freq.sum
v.to
torchvision.transforms.functional.adjust_hue
zip
torch.load
par.to
torch.arange
apex.parallel.DistributedDataParallel.fix_bn
cityscapes_cmap
self.fisher.items
torch.sum
images.to.to
dataset
argparser.get_argparser.parse_args
dataset.transform.Compose
utils.loss.BCEWithLogitsLossWithIgnoreIndex
torch.optim.lr_scheduler.StepLR
torch.nn.init.constant_
apex.parallel.DistributedDataParallel.load_state_dict
utils.logger.Logger.print
utils.logger.Logger.add_figure
dataset.transform.CenterCrop
random.randint
min
self.confusion_matrix_to_fig
ax.figure.colorbar
torch.nn.MaxPool2d
torchvision.transforms.functional.adjust_brightness
SegmentationModule
focal_loss.mean
FocalLoss
self._check_input
self.red_conv
util.try_index
images.detach.cpu
torchvision.transforms.functional.to_tensor
self.classifier
torchvision.transforms.functional.center_crop
self.map_bn
self.regularizer.penalty.item
self.cls.bias.data.copy_
torch.cuda.manual_seed
self.confusion_matrix.astype
p.detach
argparse.ArgumentParser.add_argument
str
utils.logger.Logger.debug
self.__strip_zero
list
x.view
torch.distributed.get_world_size
train.Trainer.train
outputs.max
prediction.cpu.numpy.cpu
transforms.append
self.global_pooling_conv
torch.isinf
self.logger.add_text
model.modules
torch.utils.data.random_split
self.total_samples.torch.tensor.to.cpu
mat.min
filter
train.Trainer.state_dict
ax.imshow
apex.parallel.DistributedDataParallel.to
tasks.get_per_task_classes
len
self.proj_bn
utils.get_regularizer
n.self.model_old_dict.p.pow
reg_loss.torch.tensor.to
torchvision.transforms.functional.hflip
utils.Label2Color
self.bn1
self.device.n.self.model_temp.to.p.detach.pow
torch.nn.GroupNorm
ade_cmap
Lambda
numpy.array
outputs.narrow
lbl.label2color.transpose.astype
self._std.reshape
self.model_old
torch.nn.functional.cross_entropy
RuntimeError
criterion
model
self.delta.items
self.mod1
model.state_dict
normalize_fn
params.append
par.torch.clone.to
self.logger.add_figure
self.get
models.util.try_index
ValueError
enumerate
torch.nn.Conv2d
self.mod3
self.info
model.head.parameters
img.denorm.astype
numpy.unique
model.train
n.self.score_plus_fisher.mean
random.shuffle
all
labels.remove
self.pool_red_conv
results.items
opts.backbone.models.__dict__.load_state_dict
p.to
fisher.items
inputs.shape.labels_new.F.one_hot.float.permute.clone
main
self.transform
self.book.get
format
_NETS.items
utils.logger.Logger.close
torch.sigmoid
range
dataset.transform.RandomHorizontalFlip
task_dict.keys
ax.set
self.logger.close
self.convs.clone
PI
callable
loss.mean.item
n.self.model_old_dict.p.pow.n.self.score_plus_fisher.sum
isinstance
torch.save
EWC
copy.deepcopy
torch.nn.Sequential
torch.nn.init.xavier_normal_
self.lde_loss.item
dropout
utils.Subset
self.total_samples.torch.tensor.to
self.modules
logger.add_scalar
apex.parallel.DistributedDataParallel.eval
outputs_no_bgk.labels.sum
open
self.FocalLoss.super.__init__
self.confusion_matrix.torch.tensor.to.cpu
torchvision.transforms.functional.resize
p.torch.clone.detach.cpu
utils.loss.IcarlLoss
blocks.append
self.confusion_matrix.sum
mask.float.mean
torch.manual_seed
utils.color_map
torch.nn.ModuleList
image_set.rstrip
self.IdentityResidualBlock.super.__init__
numpy.zeros
self.mod4
t.apply_
dataset.transform.ToTensor
self.global_pooling_bn
x.size.x.size.x.view.mean
self._network.append
utils.loss.UnbiasedCrossEntropy
torch.index_select
argparser.get_argparser
model.cls.parameters
torch.isnan
self._stride_dilation
cls.weight.data.copy_
torch.tensor.to
self.licarl
torchvision.transforms.functional.vflip
x.size
numpy.save
torch.cuda.set_device
utils.logger.Logger.add_scalar
numpy.diag
res.values
self.n_classes.mask.label_pred.int.mask.label_true.astype.self.n_classes.np.bincount.reshape
inputs.shape.torch.tensor.to
torch.logsumexp
collections.OrderedDict
torch.utils.data.distributed.DistributedSampler
self._network
super.__init__
torch.clone
os.listdir
FileNotFoundError
torch.where
torch.nn.functional.elu
torchvision.transforms.Lambda
bitget
new_bias.squeeze
self.confusion_matrix.torch.tensor.to
self.proj_conv
self.regularizer.state_dict
self.regularizer.penalty
metrics.StreamSegMetrics.to_str
metrics.get_results
save_ckpt
apex.parallel.DistributedDataParallel
torch.optim.SGD.state_dict
torch.device
inputs.shape.labels_new.F.one_hot.float.permute.sum
self.model_old.state_dict
mask.float.sum
max
in_size.in_size.inputs.view.mean
hasattr
logging.error
torch.mean
numpy.zeros.astype
self.add_module
ret_samples.append
focal_loss.sum
scheduler.step
apex.parallel.DistributedDataParallel.train
self.lkd_loss.item
tasks_voc.keys
dataset.transform.Normalize
n.self.score_old.to
torch.distributed.barrier
torch.distributed.init_process_group
index.self.images.Image.open.convert
functools.reduce
torch.optim.SGD
self.target_transform
torchvision.transforms.functional.normalize
confusion_matrix.cpu.numpy
target.label2color.transpose
torch.ones_like
inputs.view
model.body.parameters
self.logger.add_scalar
p.torch.clone.detach.to
super
self.reset_parameters
print
tuple
</pre>
</details>

@developer
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.
